src/transactions/CreateClaimableBalanceOpFrame.cpp

```diff
@@ -0,0 +1,274 @@
+        return pred.absAfter() >= 0;
+    case CLAIM_PREDICATE_BEFORE_RELATIVE_TIME:
+        return pred.relBefore() >= 0;
+    case CLAIM_PREDICATE_AFTER_RELATIVE_TIME:
+        return pred.relAfter() >= 0;
+    default:
+        abort();
+    }
+
+    return true;
+}
+
+CreateClaimableBalanceOpFrame::CreateClaimableBalanceOpFrame(
+    Operation const& op, OperationResult& res, TransactionFrame& parentTx,
+    size_t index)
+    : OperationFrame(op, res, parentTx)
+    , mCreateClaimableBalance(mOperation.body.createClaimableBalanceOp())
+    , mOpIndex(index)
+{
+}
+
+bool
+CreateClaimableBalanceOpFrame::isVersionSupported(
+    uint32_t protocolVersion) const
+{
+    return protocolVersion >= 14;
+}
+
+bool
+CreateClaimableBalanceOpFrame::doApply(AbstractLedgerTxn& ltx)
+{
+    auto header = ltx.loadHeader();
+    auto sourceAccount = loadSourceAccount(ltx, header);
+
+    auto const& claimants = mCreateClaimableBalance.claimants;
+
+    // Deduct claimants.size() * baseReserve of native asset
+    int64_t entryCost =
+        int64_t(header.current().baseReserve) * claimants.size();
+
+    if (getAvailableBalance(header, sourceAccount) < entryCost)
+    {
+        innerResult().code(CREATE_CLAIMABLE_BALANCE_LOW_RESERVE);
+        return false;
+    }
+
+    auto ok = addBalance(header, sourceAccount, -entryCost);
+    assert(ok);
+
+    // Deduct amount of asset from sourceAccount
+    auto const& asset = mCreateClaimableBalance.asset;
+    auto amount = mCreateClaimableBalance.amount;
+
+    if (asset.type() == ASSET_TYPE_NATIVE)
+    {
+        if (getAvailableBalance(header, sourceAccount) < amount)
+        {
+            innerResult().code(CREATE_CLAIMABLE_BALANCE_UNDERFUNDED);
+            return false;
+        }
+
+        auto ok = addBalance(header, sourceAccount, -amount);
+        assert(ok);
+    }
+    else
+    {
+        auto trustline = loadTrustLine(ltx, getSourceID(), asset);
+        if (!trustline)
+        {
+            innerResult().code(CREATE_CLAIMABLE_BALANCE_NO_TRUST);
+            return false;
+        }
+        if (!trustline.isAuthorized())
+        {
+            innerResult().code(CREATE_CLAIMABLE_BALANCE_NOT_AUTHORIZED);
+            return false;
+        }
+
+        if (!trustline.addBalance(header, -amount))
+        {
+            innerResult().code(CREATE_CLAIMABLE_BALANCE_UNDERFUNDED);
+            return false;
+        }
+    }
+
+    // Create claimable balance entry
+    LedgerEntry newClaimableBalance;
+    newClaimableBalance.data.type(CLAIMABLE_BALANCE);
+
+    auto& claimableBalanceEntry = newClaimableBalance.data.claimableBalance();
+    claimableBalanceEntry.createdBy = getSourceID();
+    claimableBalanceEntry.amount = amount;
+    claimableBalanceEntry.asset = asset;
+    claimableBalanceEntry.reserve = entryCost;
+
+    OperationID operationID;
+    operationID.type(ENVELOPE_TYPE_OP_ID);
+    operationID.id().sourceAccount = toMuxedAccount(getSourceID());
```

Comment:
I think this is wrong: you're supposed to use the sourceID from the transaction not from this operation. 

The CAP is actually confusing as it's using "sourceAccount".

Reason you need this to be the tx source account is that otherwise the operationID will not be unique (so we use `txSrcAccount | seqNum` to ensure that it's the case.

Without this change, the call to `create` below will throw with `std::runtime_error("Key already exists")`

Reply:
you should add test cases for this

Reply:
@sisuresh Can you also update the CAP to increase clarity about this?

Reply:
Good catch. Added a test case under section `validate Tx account is used in hash`. I'll also update the cap.